src/VisualStudio/Core/Def/LanguageServer/Handler/CustomMessage/CustomMessageHandler.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CustomMessage;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(CustomMessageHandler)), Shared]
+[Method(MethodName)]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class CustomMessageHandler()
+    : ILspServiceDocumentRequestHandler<CustomMessageParams, CustomResponse>
+{
+    private const string MethodName = "roslyn/customMessage";
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(CustomMessageParams request)
+    {
+        return request.Message.TextDocument;
+    }
+
+    public async Task<CustomResponse> HandleRequestAsync(CustomMessageParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        // Create the Handler instance. Requires having a parameterless constructor.
+        // ```
+        // public class CustomMessageHandler
+        // {
+        //     public Task<TResponse> ExecuteAsync(TRequest, Document, CancellationToken);
+        // }
+        // ```
+        var handler = Activator.CreateInstanceFrom(request.AssemblyPath, request.TypeFullName).Unwrap();
+
+        // Use reflection to find the ExecuteAsync method.
+        var handlerType = handler.GetType();
+        var executeMethod = handlerType.GetMethod("ExecuteAsync", BindingFlags.Public | BindingFlags.Instance);
```

Comment:
we shouldnt' use reflection like this.  We can just cast to a well known extensibility interface to call into instead.